client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,269 @@
+		let mut counter_max = 0;
+		let finalized = self.backend.blockchain().info().finalized_hash;
+
+		for leaf in self.backend.blockchain().leaves().unwrap_or_default() {
+			let route = sp_blockchain::tree_route(self.backend.blockchain(), finalized, leaf)
+				.expect("Route from finalized to leaf should be available; qed");
+			if !route.retracted().is_empty() {
+				continue
+			}
+			std::iter::once(route.common_block()).chain(route.enacted()).for_each(|elem| {
+				if !self.freshness.contains_key(&elem.hash) {
+					// Use the block height value as the freshness.
+					self.import_counter = elem.number.unique_saturated_into();
+					self.update(elem.number, elem.hash);
+				}
+			});
+			if self.import_counter > counter_max {
+				counter_max = self.import_counter;
+			}
+		}
+
+		self.import_counter = counter_max;
+		self.lowest_level = self.backend.blockchain().info().finalized_number;
+	}
+
+	/// Check and enforce the limit bound at the given height.
+	///
+	/// In practice this will enforce the given height in having a number of blocks less than
+	/// the limit passed to the constructor.
+	///
+	/// If the given level is found to have a number of blocks greater than or equal the limit
+	/// then the limit is enforced by chosing one (or more) blocks to remove.
+	/// The removal strategy is driven by the block freshness.
+	///
+	/// A block freshness is determined by the most recent leaf freshness descending from the block
+	/// itself. In other words its freshness is equal to its more "fresh" descendant.
+	///
+	/// The least "fresh" blocks are eventually removed.
+	pub fn enforce_limit(&mut self, number: NumberFor<Block>) {
+		if self.import_counter == 0 {
+			// First call detected, restore using the backend persistent information.
+			self.restore();
+		}
+
+		let level = self.levels.entry(number).or_default();
+		if level.len() < self.level_limit {
+			return
+		}
+
+		log::debug!(
+			target: "parachain",
+			"Detected leaves overflow at height {}, removing obsolete blocks", number
+		);
+
+		let mut leaves = self.backend.blockchain().leaves().unwrap_or_default();
+		// Sort leaves by freshness (most recent first).
+		leaves.sort_unstable_by(|a, b| self.freshness.get(a).cmp(&self.freshness.get(b)));
+
+		// This may not be the most efficient way to remove **multiple** entries, but is the clear
+		// one. Should be considered that in normal conditions the number of blocks to remove is 0
+		// or 1, it is not worth to complicate the code too much.
+		let remove_count = level.len() - self.level_limit + 1;
+		(0..remove_count).for_each(|_| self.remove_block(number, &leaves));
+	}
+
+	/// This will search for the less fresh block and removes it along with all its descendants.
+	/// Leaves should have already been ordered by "freshness" (older first).
+	fn remove_block(&mut self, number: NumberFor<Block>, leaves: &[Block::Hash]) {
+		let mut candidate_freshest_leaf_idx = usize::MAX;
+		let mut candidate_freshest_route = None;
```

Comment:
:-D

Let me first try to explain better the idea here:
1.Given a set of candidate blocks at the same level we want to choose one (THE target)
2. Each of these candidates may have a bunch of leaves  (if not, the candidate block itself is the leaf).
3. For each candidate we select the freshest leaf. 
    Let's say that this freshest leaf "freshness" value becomes the candidate freshness.
4. The target is chosen to be the one that is less fresh (according to 3 criteria).

Now, regarding the two variables. These are used **only** to optimize some steps.

Usage:
- `candidate_freshest_leaf_idx`: if during the loop we find another block (at the same level of the current candidate) with a leaf that is fresher (or equally fresh) than `candidate_freshest_leaf_idx` than we can early exit processing that block since for sure it is not a candidate (because it has a leaf that is fresher than the current candidate).
- `candidate_freshest_route`:  route to candidate freshest leaf. This is just to avoid getting the route to this leaf again on children removal [here]()